### PR TITLE
Improve logging on result serialization errors

### PR DESCRIPTION
**Pull request by desertaxle**
This PR updates the transaction context manager to log a warning when result serialization fails instead of a big traceback. This PR also avoids logging when the default rollback hook that updates the tasks state runs.

Very open to changing the verbiage of the warning message.

### Example output

```
15:42:51.769 | WARNING | Task run 'create_db_and_cur-5c1' - Encountered an error while serializing result for transaction '159ef706744b2d7edd395d9458f3d2ec': Failed to serialize object of type 'tuple' with serializer 'pickle'. You can try a different serializer (e.g. result_serializer="json") or disabling persistence (persist_result=False) for this flow or task. Code execution will continue, but the transaction will not be committed.
```

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

**Review by zzstoatzz — APPROVED** [review]
this makes sense to me!

File: src/prefect/states.py [hunk 1]
```diff
@@ -236,7 +242,13 @@ async def exception_to_failed_state(
     if result_factory:
         data = await result_factory.create_result(exc)
         if write_result:
-            await data.write()
+            try:
+                await data.write()
+            except Exception as exc:
+                local_logger.warning(
+                    "Failed to write result: %s Execution will continue, but the result has not been written",
```
Comment (zzstoatzz) [inline src/prefect/states.py:249]: there's a natasha bedingfield joke in here somewhere

File: src/prefect/results.py [hunk 2]
```diff
@@ -516,7 +515,11 @@ def serialize(
             bytes: the serialized record
 
         """
-        return self.model_dump_json(serialize_as_any=True).encode()
+        return (
+            self.model_copy(update={"result": self.serialize_result()})
```
[the remark this answers is not in the document]
Comment (desertaxle) [reply src/prefect/results.py:519]: Pydantic buries the `SerializationError` inside a couple of its own exceptions, so I couldn't get the original error when this was a field serializer. This is a workaround to ensure we get a clean `SerializationError` to present to the user.

File: src/prefect/states.py (hunk 1 above)
Comment (zzstoatzz) [reply src/prefect/states.py:249]: "but the result is still unwritten"

File: src/prefect/states.py [hunk 3]
```diff
@@ -343,7 +365,13 @@ async def return_value_to_state(
             expiration=expiration,
         )
         if write_result:
-            await result.write()
+            try:
+                await result.write()
+            except Exception as exc:
+                local_logger.warning(
+                    "Encountered an error while writing result: %s Execution will continue, but the result has not been written",
```
[the remark this answers is not in the document]
Comment (desertaxle) [reply src/prefect/states.py:372]: Shoot, I can update that in a new PR!

**Pull request merged**